sklearn/feature_selection/univariate_selection.py

```diff
@@ -307,16 +306,415 @@ def f_regression(X, y, center=True):
+        to continuous variables in order to remove repeated values.
+        If int, random_state is the seed used by the random number generator;
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
+    Attributes
+    ----------
+    scores_ : ndarray, shape (n_features,)
+        Estimated mutual information between each feature and the target.
+
+    Notes
+    -----
+    1. The term "discrete features" is used instead of naming them
+       "categorical", because it describes the essence more accurately.
+       For example, pixel intensities of an image are discrete features
+       (but hardly categorical) and you will get better results if mark them
+       as such. Also note, that treating a continuous variable as discrete and
+       vice versa will usually give incorrect results, so be attentive about that.
+    2. True mutual information can't be negative. If its estimate turns out
+       to be negative, it is replaced by zero.
+
+    References
+    ----------
+    .. [1] `Mutual Information <https://en.wikipedia.org/wiki/Mutual_information>`_
+           on Wikipedia.
+    .. [2] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [3] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    .. [4] L. F. Kozachenko, N. N. Leonenko, "Sample Estimate of the Entropy
+           of a Random Vector", Probl. Peredachi Inf., 23:2 (1987), 9-16
+
+    See also
+    --------
+    ANOVAFScorerClassification: ANOVA F-value between label/feature for classification tasks.
+    ANOVAFScorerRegression: ANOVA F-value between label/feature for regression tasks.
+    Chi2Scorer: Chi-squared stats of non-negative features for classification tasks.
+    MutualInfoScorerClassification: Mutual information for a discrete target.
+    """
+    def __init__(self, discrete_features='auto', n_neighbors=3,
+                 copy=True, random_state=None):
+        self.discrete_features = discrete_features
+        self.n_neighbors = n_neighbors
+        self.copy = copy
+        self.random_state = random_state
+
+    def fit(self, X, y):
+        """Run scorer on (X, y).
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        self.scores_ = mutual_info_regression(X, y, self.discrete_features,
+                                              self.n_neighbors, self.copy,
+                                              self.random_state)
+        return self
+
+    def _check_params(self, X, y):
+        pass
+
+
```

Comment:
I agree that this is a lot of unnecessary code. I haven't checked the whole PR, but at least for this part of the code, another design approach is to have a very small `create_scorer_class` or something, which returns an instance of the class you need in for the purpose you need it. I see no point in having a whole public class with functions doing nothing (`pass`).

Reply:
I think you are looking at things a bit harshly.  If what you highlighted is not necessary it can removed, I wasn't sure with the sklearn framework if certain placeholder methods are necessary or not.  We can remove all pass methods that aren't needed, just tell me.

The question you have to ask is if we could have a clean slate with scoring functions, would you make them proper classes and estimators?  My answer to that is unequivocally yes.

The reason I had to make the scorer classes as wrappers to the functions is to maintain backward compatibility and not duplicate code.  The way I should've redesigned it is to move all the scoring function code into the fit() methods and then made wrappers for the legacy functions to maintain backward compatibility, then it would be more clear and compelling.